sqllogictest/src/connection.rs

```diff
@@ -0,0 +1,90 @@
+use std::collections::HashMap;
+use std::future::IntoFuture;
+
+use futures::Future;
+
+use crate::{AsyncDB, Connection as ConnectionName, DBOutput};
+
+/// Trait for making connections to an [`AsyncDB`].
+///
+/// This is introduced to allow querying the database with different connections
+/// (then generally different sessions) in a single test file with `connection` records.
+pub trait MakeConnection {
+    /// The database type.
+    type Conn: AsyncDB;
+    /// The future returned by [`MakeConnection::make`].
+    type MakeFuture: Future<Output = Result<Self::Conn, <Self::Conn as AsyncDB>::Error>>;
+
+    /// Creates a new connection to the database.
+    fn make(&mut self) -> Self::MakeFuture;
+}
+
+/// Make connections directly from a closure returning a future.
+impl<D: AsyncDB, F, Fut> MakeConnection for F
+where
+    F: FnMut() -> Fut,
+    Fut: IntoFuture<Output = Result<D, D::Error>>,
+{
+    type Conn = D;
+    type MakeFuture = Fut::IntoFuture;
+
+    fn make(&mut self) -> Self::MakeFuture {
+        self().into_future()
+    }
+}
+
+/// Make connections with a synchronous infallible function.
+#[derive(Debug)]
+pub struct MakeWith<F>(pub F);
+
+impl<F: FnMut() -> D, D: AsyncDB> MakeConnection for MakeWith<F> {
+    type Conn = D;
+    type MakeFuture = futures::future::Ready<Result<D, D::Error>>;
+
+    fn make(&mut self) -> Self::MakeFuture {
+        futures::future::ready(Ok((self.0)()))
+    }
+}
+
+/// Connections established in a [`Runner`](crate::Runner).
+pub(crate) struct Connections<M: MakeConnection> {
+    make_conn: M,
+    conns: HashMap<ConnectionName, M::Conn>,
+}
+
+impl<M: MakeConnection> Connections<M> {
```

Comment:
nit: guess `impl<D: AsyncDB, M: MakeConnection<Conn = D>>` can make it slightly cleaner?

Reply:
I'll have a try.

Reply:
Wow, this looks nice!